c7n/utils.py

```diff
@@ -818,20 +818,31 @@ def merge_dict_list(dict_iter):
 
 
 def merge_dict(a, b):
-    """Perform a merge of dictionaries a and b
+    """Perform a merge of dictionaries A and B
 
     Any subdictionaries will be recursively merged.
-    Any leaf elements in the form of a list or scalar will use the value from a
+    Any leaf elements in the form of scalar will use the value from A
+    If there is a scalar and list for the same key, the scalar will be appended to the list
+    If there are two lists for the same key, the list from B will be treated like a set
+    and be appended to the list from A
     """
-    d = {}
-    for k, v in a.items():
-        if k not in b:
-            d[k] = v
-        elif isinstance(v, dict) and isinstance(b[k], dict):
-            d[k] = merge_dict(v, b[k])
+    d = a.copy()
```

Comment:
Good thinking, this seems to more clearly show the "we prioritize vals from A"` :+1: . It also smells a bit more efficient at a glance, though given the potential for deep recursive merges we'd probably want a [deepcopy](https://docs.python.org/3/library/copy.html#copy.deepcopy). The prior logic seemed to be spending a bit more effort on checks up front to avoid some wasted copying.

Reply:
Thanks, it makes more sense to use deepcopy here.